{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import defopt\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.metrics import confusion_matrix, ConfusionMatrixDisplay\n",
    "from tqdm import tqdm\n",
    "import time\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import re\n",
    "import scipy.sparse as ss\n",
    "import h5py\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_genes(gene_names):\n",
    "    # get rid of gene models etc\n",
    "    genes_Rik = np.array([ re.search('Rik$', s) is not None for s in gene_names ])\n",
    "    genes_Gm = np.array([ re.search('Gm\\d', s) is not None for s in gene_names ])\n",
    "    genes_LOC = np.array([ re.search('LOC\\d', s) is not None for s in gene_names ])\n",
    "    genes_AA = np.array([ re.search('^[A-Z]{2}\\d*$', s) is not None for s in gene_names ])\n",
    "    keep_genes = np.logical_not(genes_Rik + genes_Gm + genes_LOC + genes_AA)\n",
    "    \n",
    "    #Pseudocode for filtering out genes that are below a mean read threshold across all clusters\n",
    "    #genes_lowcount = np.array()\n",
    "    # _, _, cluster_means, _ = compute_means(load_data_tasic_2018('c:\\\\Users\\\\Alex\\\\PhD_data\\\\allen2018'), 'cluster') # n clusters x n genes matrix\n",
    "    # gene_threshold = 0\n",
    "    # for gene in gene_names:\n",
    "    #     if np.max(cluster_means[gene] < gene_threshold):\n",
    "    #         genes_lowcount.append(gene)\n",
    "    keep_genes = np.logical_not(genes_Rik + genes_Gm + genes_LOC + genes_AA)\n",
    "    return keep_genes\n",
    "\n",
    "def load_data_tasic_2018(datapath, filter_neurons=True):\n",
    "    \"\"\"\n",
    "    Load the scRNAseq data from Tasic et al., \"Shared and distinct\n",
    "    transcriptomic cell types across neocortical areas\", Nature, 2018.\n",
    "\n",
    "    Args:\n",
    "        datapath: path to the data\n",
    "\n",
    "    Returns:\n",
    "        exons_matrix: n cells x n genes matrix (numpy.ndarray) of read counts\n",
    "        cluster_ids: numpy.array of cluster assignments from the cell metadata\n",
    "        cluster_means: n clusters x n genes matrix (numpy.ndarray) of mean read\n",
    "            counts for each cluster\n",
    "        cluster_labels: list of cluster names\n",
    "        gene_names: pandas.Series of gene names\n",
    "\n",
    "    \"\"\"\n",
    "    fname_metadata = f'{datapath}mouse_VISp_2018-06-14_samples-columns.csv'\n",
    "    metadata = pd.read_csv(fname_metadata, low_memory=False)\n",
    "    fname = f'{datapath}mouse_VISp_2018-06-14_exon-matrix.csv'\n",
    "    exons = pd.read_csv(fname, low_memory=False)\n",
    "    fname_genes = f'{datapath}mouse_VISp_2018-06-14_genes-rows.csv'\n",
    "    genes = pd.read_csv(fname_genes, low_memory=False)\n",
    "    gene_names = genes['gene_symbol']\n",
    "    metadata.set_index('sample_name', inplace=True)\n",
    "    keep_genes = filter_genes(gene_names)\n",
    "    exons = exons.iloc[keep_genes]\n",
    "    gene_names = gene_names.iloc[keep_genes]\n",
    "    exons_df = metadata.join(exons.T, on='sample_name')\n",
    "    # only include neurons\n",
    "    include_classes = [\n",
    "        'GABAergic',\n",
    "        'Glutamatergic'\n",
    "    ]\n",
    "    # get rid of low quality cells etc\n",
    "    if filter_neurons:\n",
    "        exons_subset = exons_df[exons_df['class'].isin(include_classes)]\n",
    "    else:\n",
    "        exons_subset = exons_df\n",
    "    exons_subset = exons_subset[~exons_subset['cluster'].str.contains('ALM')]\n",
    "    exons_subset = exons_subset[~exons_subset['cluster'].str.contains('Doublet')]\n",
    "    exons_subset = exons_subset[~exons_subset['cluster'].str.contains('Batch')]\n",
    "    exons_subset = exons_subset[~exons_subset['cluster'].str.contains('Low Quality')]\n",
    "    exons_subset = exons_subset[~exons_subset['subclass'].str.contains('High Intronic')]\n",
    "\n",
    "    return exons_subset, gene_names\n",
    "\n",
    "def compute_means(exons_df, classify_by, gene_filter='\\d'):\n",
    "    exons_matrix = exons_df.filter(regex=gene_filter).to_numpy() # n cells x n genes matrix\n",
    "    # names of columns containing expression data are integer numbers\n",
    "    expression_by_cluster = exons_df.groupby([classify_by]).mean().filter(regex=gene_filter)\n",
    "    cluster_means = expression_by_cluster.to_numpy()   # n clusters x n genes matrix\n",
    "\n",
    "    cluster_ids = np.empty(len(exons_df[classify_by]), dtype=int)\n",
    "    for i, cluster in enumerate(exons_df[classify_by]):\n",
    "        cluster_ids[i] = np.nonzero(expression_by_cluster.index == cluster)[0]\n",
    "    cluster_labels = expression_by_cluster.index\n",
    "    return exons_matrix, cluster_ids, cluster_means, cluster_labels\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# n clusters x n genes matrix\n",
    "exons_subset, gene_names = load_data_tasic_2018('C:/Users/Alex/PhD_data/allen2018/', filter_neurons=True)\n",
    "exons_matrix, cluster_ids, cluster_means, cluster_labels = compute_means(exons_subset, 'cluster')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "max_cluster_means = []\n",
    "count = -1\n",
    "for array in cluster_means:\n",
    "    for gene in array:\n",
    "        count +=1\n",
    "        max_cluster_means.append(max(cluster_means[:,count]))\n",
    "    if count > 23198:\n",
    "            break\n",
    "\n",
    "genes = gene_names.tolist()\n",
    "gene_counts = np.column_stack((genes, max_cluster_means))\n",
    "\n",
    "np.savetxt(\"C:/Users/Alex/PhD_data/allen2018/cluster_means.csv\", gene_counts, delimiter=\",\", fmt = '%s')\n",
    "\n",
    "filtered_gene_counts = np.delete(gene_counts, np.where(gene_counts[:,1].astype('float64') == 0), axis=0)\n",
    "padlocks_required = np.column_stack((filtered_gene_counts[:,0], 38748 / filtered_gene_counts[:,1].astype('float64')))\n",
    "np.savetxt(\"C:/Users/Alex/PhD_data/allen2018/padlocks_required.csv\", padlocks_required, delimiter=\",\", fmt = '%s')\n",
    "\n",
    "filtered_padlocks_required = np.delete(padlocks_required, np.where(padlocks_required[:,1].astype('float64') > 100), axis=0)\n",
    "np.savetxt(\"C:/Users/Alex/PhD_data/allen2018/filtered_padlocks_required.csv\", filtered_padlocks_required, delimiter=\",\", fmt = '%s')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "picked_genes = ['Aqp4', 'Arpp21', 'Brinp3', 'Calb1', 'Calb2', 'Cartpt', 'Cck', 'Cd24a',\n",
    "       'Cdh13', 'Chn2', 'Chodl', 'Chrm2', 'Cnr1', 'Cplx3', 'Cpne6', 'Crh', 'Cryab',\n",
    "       'Cxcl12', 'Cxcl14', 'Dkk3', 'Enpp2', 'Etv1', 'Fst', 'Gabra1',\n",
    "       'Gad1', 'Gap43', 'Gpx3', 'Hpcal4', 'Id2', 'Igfbp4', 'Itm2c',\n",
    "       'Kcnab1', 'Lamp5', 'Lypd6', 'Lypd6b', 'Marcksl1', 'Mdh1', 'Myl4',\n",
    "       'Ncald', 'Nefl', 'Nnat', 'Nov', 'Npy', 'Nr4a2', 'Nrep', 'Nrsn1',\n",
    "       'Olfm3', 'Pantr1', 'Pcdh8', 'Pcp4', 'Pcp4l1', 'Pde1a', 'Pdyn',\n",
    "       'Penk', 'Prss23', 'Ptn', 'Pvalb', 'Rab3b', 'Rbp4', 'Rcan2', 'Reln',\n",
    "       'Rgs10', 'Rgs4', 'Rspo1', 'Scg2', 'Serpine2', 'Serpini1','Slc1a3', 'Snca',\n",
    "       'Sncg', 'Sparcl1', 'Spock3', 'Spon1', 'Sst', 'Stmn2', 'Stxbp6', 'Synpr',\n",
    "       'Tac2', 'Thsd7a', 'Vip']\n",
    "\n",
    "picked_gene_counts = []\n",
    "for gene in picked_genes:\n",
    "    picked_gene_counts.append(gene_counts[int(np.where(gene_counts == gene)[0]),1])\n",
    "\n",
    "\n",
    "picked_gene_counts = np.column_stack((picked_genes, picked_gene_counts))\n",
    "np.savetxt(\"C:/Users/Alex/PhD_data/allen2018/picked_gene_counts.csv\", picked_gene_counts, delimiter=\",\", fmt = '%s')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "barseq_genes = ['Calb1','Car3', 'Ccn2','Cdh10','Cdh11','Cdh12','Cdh13','Cdh15','Cdh18','Cdh20','Cdh22','Cdh24','Cdh4','Cdh6','Cdh8','Cdh9','Col11a1',\n",
    "    'Col19a1','Coro6','Cplx3','Enpp2','Fbxl7','Fezf2','Foxp2','Gad1','Galnt14','Galntl6','Gfra1','Grik1','Hgf','Hs3st4','Il1rapl2','Inpp4b','Kcnip1','Kcnn2','Lpp','Nr4a2','Nrp1','Nxph1','Oprk1','Otof','Pcdh11x','Pcdh17',\n",
    "    'Pcdh19','Pcdh20','Pcdh7','Pcdh8','Pcdh9','Rab3c','Rasgrf2','Rasl10a','Rorb','Sdk1','Slc17a7','Slc24a3','Slc30a3','Sorcs3','Svil','Synpr','Tafa1','Tafa2','Tle4','Tshz2','Vwc2l','Zfpm2']\n",
    "\n",
    "barseq_picked_gene_counts = []\n",
    "\n",
    "for gene in barseq_genes:\n",
    "    barseq_picked_gene_counts.append(gene_counts[int(np.where(gene_counts == gene)[0]),1])\n",
    "\n",
    "\n",
    "barseq_picked_gene_counts = np.column_stack((barseq_genes, barseq_picked_gene_counts))\n",
    "np.savetxt(\"C:/Users/Alex/PhD_data/allen2018/barseq_picked_gene_counts.csv\", barseq_picked_gene_counts, delimiter=\",\", fmt = '%s')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "4de1947290903876fcddc25d8bf621ee8bd7e394cd575d7648578f103bce38d9"
  },
  "kernelspec": {
   "language": "python"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
